agent/tools.py

```
@@ -118,6 +118,66 @@ def apply_diff(self, path: str, new_content: str) -> dict[str, str]:
+        if not target.exists():
+            return {"path": path, "diff": "", "error": f"file not found: {path!r}"}
+
+        original = target.read_text(encoding="utf-8")
+
+        # Exact match first.
+        if old_string in original:
+            updated = original.replace(old_string, new_string, 1)
+        else:
+            # Whitespace-normalised fallback: strip trailing spaces per line so
+            # that copy-paste differences (e.g. trailing tabs) don't block the edit.
+            stripped_original = "\n".join(l.rstrip() for l in original.splitlines())
```

Ruff (0.16.1)

[error] Ambiguous variable name: `l`

(E741)